### PR TITLE
Implementation of the Language Server Supervisor Process

### PR DESCRIPTION
**Pull request by lolczak**
### Pull Request Description
A process that sits in between the client and the language server to restart it if it dies.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: doc/design/engine/engine-services.md [hunk 1]
```diff
@@ -1887,6 +1889,33 @@ null
...
+### Monitoring
+The language server also has a heartbeat operation to monitor the Language
+server. This API is private and should be used only by the Project Manager.
+
+#### `heartbeat/ping`
+This request is sent from the supervisor process to the server to check the
+health of the Language Server.
+
+- **Type:** Request
+- **Direction:** Supervisor -> Server
```
Comment (iamrecursion) [inline doc/design/engine/engine-services.md:1901]:
```suggestion
- **Direction:** Supervisor -> Server
- **Connection:** Protocol
- **Visibility:** Private
```

Just so it's up to date with my branch's formatting.

File: common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerSupervisor.scala [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+          connectionFactory,
+          scheduler
+        )
+      )
+
+    case ServerUnresponsive =>
+      log.info(s"Server is unresponsive [$config]. Restarting it...")
+      cancellable.cancel()
+      log.info(s"Restarting first time the server")
+      server.restart() pipeTo self
```
Comment (4e6) [inline common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerSupervisor.scala:75]: Just a comment on a Heartbeat implementation. 
Networks are unreliable, and a single heartbeat timeout can be a false positive. Ideally, we want to wait for a couple of missed heartbeats to make sure it is not a network lag.
On the other hand, waiting for extra heartbeats can decrease the responsiveness of service. So if the restart operation is cheap, frequent restarts is probably a better choice.
I can't decide which behavior we want here.

Comment (lolczak) [reply common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerSupervisor.scala:75]: I agree with you that it is an unreliable test for services interconnected with networking equipment.  Nevertheless, a process manager and language servers will communicate using the loopback interface, and all communication will be routed through the kernel. Thus, we can assume that the communication channel is reliable.

Comment (iamrecursion) [reply common/project-manager/src/main/scala/org/enso/projectmanager/infrastructure/languageserver/LanguageServerSupervisor.scala:75]: Just to weigh in, for simplicity's sake, I think we should stick with assuming reliability of the channel for now. If it proves to be a problem we can change things later.

**Pull request merged**